src/lib.rs

```diff
@@ -621,182 +654,203 @@ where
-        {
-            let name = Self::resolve_name(&specification, header, debug_str, abbrev)
-                .chain_err(|| "specification does not resolve to a name")?;
-            return Ok(name);
+
+        if let &Some(ref specification) = specification {
+            if let Some(name) = self.resolve_indirect_name(specification)
+                .chain_err(|| "invalid subprogram specification origin")?
+            {
+                return Ok(Some(name));
+            }
         }
 
         Ok(None)
     }
 
-    fn get_entry<'a>(
-        entry: &gimli::DebuggingInformationEntry<'a, 'a, gimli::EndianBuf<'input, Endian>>,
-        header: &'a gimli::CompilationUnitHeader<gimli::EndianBuf<'input, Endian>>,
-        abbrev: &'a gimli::Abbreviations,
-        attr: gimli::DwAt,
-    ) -> Result<Option<gimli::DebuggingInformationEntry<'a, 'a, gimli::EndianBuf<'input, Endian>>>> {
-        if let Some(gimli::AttributeValue::UnitRef(offset)) = entry
-            .attr_value(attr)
-            .map_err(|e| Error::from(ErrorKind::Gimli(e)))?
-        {
-            let mut entries = header.entries_at_offset(abbrev, offset)?;
-            let (_, entry) = entries.next_dfs()?.ok_or_else(|| {
-                ErrorKind::InvalidDebugSymbols(DebugInfoError::DanglingEntryOffset)
-            })?;
-            return Ok(Some(entry.clone()));
-        }
+    fn resolve_indirect_name(
+        &self,
+        attr: &gimli::AttributeValue<gimli::EndianBuf<'input, Endian>>,
+    ) -> Result<Option<gimli::EndianBuf<'input, Endian>>> {
+        let mut entries = match attr {
+            &gimli::AttributeValue::UnitRef(offset) => {
+                self.header.entries_at_offset(&self.abbrev, offset)?
+            }
+            // FIXME: handle AttributeValue::DebugInfoRef
+            _ => return Ok(None),
+        };
+        let (_, entry) = entries.next_dfs()?.ok_or_else(|| {
+            ErrorKind::InvalidDebugSymbols(DebugInfoError::DanglingEntryOffset)
+        })?;
 
-        Ok(None)
-    }
+        let mut name = None;
+        let mut abstract_origin = None;
+        let mut specification = None;
 
-    fn get_ranges(
-        entry: &gimli::DebuggingInformationEntry<gimli::EndianBuf<Endian>>,
-        debug_ranges: &gimli::DebugRanges<gimli::EndianBuf<Endian>>,
-        address_size: u8,
-        base_address: u64,
-    ) -> Result<Vec<gimli::Range>> {
-        if let Some(range) =
-            Self::parse_noncontiguous_ranges(entry, debug_ranges, address_size, base_address)?
+        let mut attrs = entry.attrs();
+        while let Some(attr) = attrs
+            .next()
+            .map_err(|e| Error::from(ErrorKind::Gimli(e)))
+            .chain_err(|| "invalid subprogram attribute")?
         {
-            return Ok(range);
-        }
-        if let Some(range) = Self::parse_contiguous_range(entry)?.map(|range| vec![range]) {
-            return Ok(range);
+            match attr.name() {
```

Comment:
This is essentially the same name matching as in `resolve_name`. Should be factored out.

Reply:
See reply to comment in `resolve_name`.